### PR TITLE
Fix term stats when talking to ES 6

### LINKED ISSUE #75349: dfs_query_then_fetch can cause serialization errors with 6.x nodes
In a mixed 6.x and 7.x cluster, a search that uses `dfs_query_then_fetch` can cause a transport serialization error:

```
[2021-07-14T14:48:54,223][DEBUG][o.e.a.s.TransportSearchAction] [v6.8.11-2] [16] Failed to execute query phase
 org.elasticsearch.transport.RemoteTransportException: [v6.8.11-0][127.0.0.1:50319][indices:data/read/search[phase/query/id]]
Caused by: java.lang.IllegalArgumentException: totalTermFreq must be positive, totalTermFreq: -1
      at org.apache.lucene.search.TermStatistics.<init>(TermStatistics.java:70) ~[lucene-core-7.7.3.jar:7.7.3 1a0d2a901dfec93676b0fe8be425101ceb754b85 - noble - 2020-04-21 10:31:55]
      at org.elasticsearch.search.dfs.AggregatedDfs.<init>(AggregatedDfs.java:37) ~[elasticsearch-6.8.11.jar:6.8.11]
      at org.elasticsearch.search.query.QuerySearchRequest.<init>(QuerySearchRequest.java:48) ~[elasticsearch-6.8.11.jar:6.8.11]
      at org.elasticsearch.transport.RequestHandlerRegistry.newRequest(RequestHandlerRegistry.java:48) ~[elasticsearch-6.8.
```

This seems related to https://issues.apache.org/jira/browse/LUCENE-8007, which was introduced in Lucene 8 and adds stricter checks to `TermStatistics`.

I was able to reproduce this with our rolling upgrade tests. A rough example:
1. In the old 6.8 cluster, create a couple documents with a `keyword` value like `"field": "some_value"`.
2. In a mixed 6.8 and 7.x cluster, perform a search on the field:
    ```
    GET /index/_search?search_type=dfs_query_then_fetch
    {
      "query": {
        "match": {
          "field": {
            "query": "some_value"
          }
        }
      }
    }
    ```

This search can then fail with the same serialization error.

### PR DESCRIPTION
**Pull request by ywelsch**
In a mixed 6.x and 7.x cluster, a search that uses dfs_query_then_fetch can cause a transport serialization errors. 

This is related to https://issues.apache.org/jira/browse/LUCENE-8007, which was introduced in Lucene 8 and adds stricter checks to TermStatistics and CollectionStatistics, and https://issues.apache.org/jira/browse/LUCENE-8020, which was introduced in Lucene 8 and avoids bogus term stats (e.g. docfreq=0). 

Co-authored-by: Julie Tibshirani <julietibs@gmail.com>

Closes #75349

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/search/dfs/DfsSearchResult.java [hunk 1]
```diff
@@ -168,6 +168,23 @@ public  static void writeSingleTermStats(StreamOutput out, TermStatistics termSt
...
                 sumTotalTermFreq = subOne(in.readVLong());
                 sumDocFreq = subOne(in.readVLong());
+                if (sumTotalTermFreq == -1L) {
+                    // fallback used by Lucene in ES 6 (LUCENE-8007)
+                    sumTotalTermFreq = sumDocFreq;
+                }
+                if (maxDoc == 0L) {
+                    // fallback used by Lucene in ES 6 (LUCENE-8007)
+                    maxDoc = docCount;
+                }
```
Comment (jpountz) [inline server/src/main/java/org/elasticsearch/search/dfs/DfsSearchResult.java:178]: Lucene seems to be doing it the other way arround: `if (docCount == -1) { docCount = maxDoc; }` reading https://github.com/apache/lucene-solr/blob/releases/lucene-solr/7.7.3/lucene/core/src/java/org/apache/lucene/search/similarities/BM25Similarity.java#L99.

File: server/src/main/java/org/elasticsearch/search/dfs/AggregatedDfs.java [hunk 2]
```diff
@@ -32,9 +34,17 @@ public AggregatedDfs(StreamInput in) throws IOException {
         termStatistics = HppcMaps.newMap(size);
         for (int i = 0; i < size; i++) {
             Term term = new Term(in.readString(), in.readBytesRef());
-            TermStatistics stats = new TermStatistics(in.readBytesRef(),
-                in.readVLong(),
-                DfsSearchResult.subOne(in.readVLong()));
+            BytesRef term2 = in.readBytesRef();
+            final long docFreq = in.readVLong();
+            assert docFreq >= 0;
+            long totalTermFreq = DfsSearchResult.subOne(in.readVLong());
```
Comment (jtibshirani) [inline server/src/main/java/org/elasticsearch/search/dfs/AggregatedDfs.java:40]: It'd be great to add a comment here explaining why this backwards compatibility logic is needed. I also wonder if it's worth factoring out a common method like `readSingleTermStats` to use in both `AggregatedDfs` and `DfsSearchResult`?

**Review by jtibshirani — COMMENTED** [review]
Thanks for picking this up!

File: server/src/main/java/org/elasticsearch/search/dfs/DfsSearchResult.java (hunk 1 above)
Comment (ywelsch) [reply server/src/main/java/org/elasticsearch/search/dfs/DfsSearchResult.java:178]: Makes sense. I'm mainly wondering how to exercise this logic (I can't see a way for this to happen with the codecs that we offer).

Comment (jpountz) [reply server/src/main/java/org/elasticsearch/search/dfs/DfsSearchResult.java:178]: Right, the default codec would always return a non-negative docCount. Maybe one approach would be to have this case only covered through a unit test that emulates 6.x serialization logic.

File: server/src/main/java/org/elasticsearch/search/dfs/AggregatedDfs.java (hunk 2 above)
Comment (ywelsch) [reply server/src/main/java/org/elasticsearch/search/dfs/AggregatedDfs.java:40]: > I also wonder if it's worth factoring out a common method like readSingleTermStats to use in both AggregatedDfs and DfsSearchResult

These methods are similar, but not quite the same. I'm therefore inclined to keep them separate.

> It'd be great to add a comment here explaining why this backwards compatibility logic is needed.

I have added docs to that effect in 0aa3cf78f08.

File: server/src/main/java/org/elasticsearch/search/dfs/DfsSearchResult.java (hunk 1 above)
Comment (ywelsch) [reply server/src/main/java/org/elasticsearch/search/dfs/DfsSearchResult.java:181]: I've added a unit test in 5abe7d351d9. Not sure it adds much though.

**Review by jpountz — APPROVED** [review]

**Review by jtibshirani — APPROVED** [review]

**Pull request merged**